### PR TITLE
Support AMP in nn.parallel

### PR DESCRIPTION
**Pull request by hfxunlp**
Take care of the state of autocast in `parallel_apply`, so there is no need to decorate model implementations.

### CODE REVIEW

**Review by mrshenli — COMMENTED** [review]
Hey @anoidgit, thanks for adding this. Shall we add a test to cover the new code?

**Comment by anoidgit** [comment]
@mrshenli Hi, with pleasure. It would be better if there was a corresponding test, but I do not know how to make it, I cannot also find the test code for `nn.parallel` for reference. If it is necessary, I hope you or someone else who are good at this can make it. Anyway, I think the code shall be fine, I have used the after modification version for my experiments.

**Comment by mrshenli** [comment]
Hey @anoidgit, you can add this to `test_data_parallel.py`. See the example below:

https://github.com/pytorch/pytorch/blob/ccc831ae358278a0699df9aea236234b42bf2180/test/distributed/test_data_parallel.py#L80-L96

**Comment by anoidgit** [comment]
Hi @mrshenli , thanks a lot for your help. I made the test following the example, please help ensure its correctness.

File: test/distributed/test_data_parallel.py [hunk 1]
```diff
@@ -95,6 +96,26 @@ def test_parallel_apply(self):
...
+        i1 = torch.randn(2, 10, device="cuda:0", dtype=torch.float)
+        i2 = torch.randn(2, 10, device="cuda:1", dtype=torch.float)
+        with autocast():
+            expected1 = l1(i1)
+            expected2 = l2(i2)
+        modules = (l1, l2)
+        expected_outputs = (expected1, expected2)
+
+        # each input can be either a collection of positional arguments
+        #                       or an object representing the single argument
```
Comment (mrshenli) [inline test/distributed/test_data_parallel.py:112]: any reason for the long indent before "or"?

**Review by mrshenli — APPROVED** [review]
LGTM! Just a minor comment. Thanks for contributing!

**Comment by mcarilli** [comment]
Code changes look good, thanks!  [DP + autocast documentation](https://pytorch.org/docs/stable/notes/amp_examples.html#dataparallel-in-a-single-process) should also be updated, but I can do that myself in a separate PR.

**Comment by anoidgit** [comment]
@mcarilli many thanks for your efforts :)

File: test/distributed/test_data_parallel.py (hunk 1 above)
Comment (anoidgit) [reply test/distributed/test_data_parallel.py:112]: thanks for pointing this out. It also looks strange to me, but I do not know the reason, I am following [the example](https://github.com/pytorch/pytorch/blob/ccc831ae358278a0699df9aea236234b42bf2180/test/distributed/test_data_parallel.py#L91-L92). If it is not right, please help fix.

Comment (mrshenli) [reply test/distributed/test_data_parallel.py:112]: that might be some debug vestige. No worries, I will land this as is, and submit a PR to fix both.

**Pull request closed without merging**